File system race condition when loading tray file

Tray files represent items to display in the system tray. They are read and parsed for information on what to display. It is possible for the tray files to be modified while the read occurs, leading to a loading error.

Expected Behavior
-----------------

After a tray file is written to, it should be read and the relevant items on the system tray should change to reflect this. Phile should be able to handle a tray file being modified.

Current Behavior
----------------

If the modifications occur in such a way to affect the tray file parsing, then exceptions might be raised.

Possible Solution
-----------------

There are different ways to handle this. One is to handle the exception by abandoning the read. If the file is being modified, Phile would be notified again later after the modification completes, and the update can be attempted again.

An alternative is to require cooperative locking before read and write. However, this removes the convenience of displaying tray icons by creating simple files.

Caching the entire tray file content when reading and before writing may help to reduce its occurrence.

Not responding to create-file events may also help to reduce its occurrence since they represent file nodes being added without any content.

Steps to Reproduce
------------------

Since this is a race condition on the file system level, it is difficult to reproduce this. I have yet to be able to reproduce it.

Context (Environment)
---------------------

Travis CI [build](https://travis-ci.com/github/BoniLindsley/phile/builds/188453400) had passed the unit tests for Python 3.6 and Python 3.7. However they failed in job [#13.3](https://travis-ci.com/github/BoniLindsley/phile/jobs/396054199) targetting Python 3.8. Relevant lines from the log of the failing job:

    ........................Exception in thread Thread-103:
    Traceback (most recent call last):
    """ Trimmed """
      File "/home/travis/build/BoniLindsley/phile/phile/tray/tmux.py", line 557, in load
        tray_file.load()
      File "/home/travis/build/BoniLindsley/phile/phile/tray/gui.py", line 124, in load
        json_content = json.load(file_stream)
    """ Trimmed """
      File "/opt/python/3.8.0/lib/python3.8/json/decoder.py", line 340, in decode
        raise JSONDecodeError("Extra data", s, end)
    json.decoder.JSONDecodeError: Extra data: line 1 column 3 (char 2)
    F.........
    ======================================================================
    FAIL: test_file_changes_after_show_and_then_hide (test_tray.test_tmux.TestIconList)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/travis/build/BoniLindsley/phile/test_phile/test_tray/test_tmux.py", line 607, in test_file_changes_after_show_and_then_hide
        self.control_mode.send_command.assert_called_with(
      File "/opt/python/3.8.0/lib/python3.8/unittest/mock.py", line 907, in assert_called_with
        raise AssertionError(_error_message()) from cause
    AssertionError: expected call not found.
    Expected: send_command('set-option -g status-right 12/2345')
    Actual: send_command('set-option -g status-right 2345')

The latter exception was raised from [this unit test](https://github.com/BoniLindsley/phile/blob/6de42d3bb0337dca1c8295a73f45218304f6f3dc/test_phile/test_tray/test_tmux.py#L607). The exception comes from a timeout, from waiting for a command to be sent from a different thread, expecting it to prepend `12/` to the `status-right` line. The thread that should have sent the command had instead raised an `JSONDecodeError` while [loading](https://github.com/BoniLindsley/phile/blob/6de42d3bb0337dca1c8295a73f45218304f6f3dc/phile/tray/gui.py#L124) the tray file containing the text `12/` to append, and so it was unable to send the command.

Detailed Description
--------------------

What I think had happened during the [unit test](https://github.com/BoniLindsley/phile/blob/6de42d3bb0337dca1c8295a73f45218304f6f3dc/test_phile/test_tray/test_tmux.py#L607):

1.  The unit test created a tray file with some content, and waited for the monitoring thread to detect the file.

    ```py
    _logger.debug('Adding a second tray file.')
    month_tray_file = TrayFile(
        configuration=self.configuration, name='month'
    )
    month_tray_file.text_icon = '12/'
    month_tray_file.save()
    expected_status_right = CommandBuilder.set_global_status_right(
        month_tray_file.text_icon + year_tray_file.text_icon
    )
    while event_setter.wait(wait_time.total_seconds()):
       """ Trimmed """
    ```py

    Saving to a new file occured in two steps. It first created the file, empty, and then wrote the content to it. This produced two file system events. A creation event, and then a modification event.

2.  While the unit test was still in the `save()` method, and created the file to be written to, the monitoring thread detected the creation event, and tried to parse it. In a tray file, the first line represents texts to display in a text-based system tray. The remaining lines form a JSON object describing other properties of the system tray. The [loading](https://github.com/BoniLindsley/phile/blob/6de42d3bb0337dca1c8295a73f45218304f6f3dc/phile/tray/gui.py#L124) read the first line, and then decoded the rest as a JSON string:

    ```py
    with self.path.open() as file_stream:

         self.text_icon = file_stream.readline().rstrip('\r\n')
         _logger.debug("Using text icon: %s", self.text_icon)

         # Make sure there are content to read by reading one byte
         # and then resetting the offset before decoding.
         current_offset = file_stream.tell()
         if file_stream.read(1):
             file_stream.seek(current_offset)
             json_content = json.load(file_stream)
         else:
             json_content = {}
    ```

    So the monitoring thread tried to read the first line in `readline()`. It found nothing because the file was just created. Note that the tray file reading stream remains at position zero at the start of the stream.

3.  While the unit test was reading or finished reading the first line due to the creation event, the unit test thread wrote the tray file content, still in the `save()` method.

4.  At this point, the monitoring thread continued to load the tray file.  It tested for content to read with `read(1)`. There was content, so `seek()` moved the read position back, and used `json.load()` to parse the rest of the file as a JSON string.

    However this started from the start of the file at position zero instead of the second line, as it normally should have done. Since the tray file was just written to, the first line contained `12/`. Trying to decode this raises a `JSONDecodeError`.

    ```
    $ python3
    Python 3.8.6 (default, Sep 25 2020, 09:36:53)
    [GCC 10.2.0] on linux
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import io, json
    >>> json.load(io.StringIO('12/'))
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/usr/lib/python3.8/json/__init__.py", line 293, in load
        return loads(fp.read(),
      File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
        return _default_decoder.decode(s)
      File "/usr/lib/python3.8/json/decoder.py", line 340, in decode
        raise JSONDecodeError("Extra data", s, end)
    json.decoder.JSONDecodeError: Extra data: line 1 column 3 (char 2)
    >>>
    ```

    With an unhandled exception, the monitoring thread then stopped.

5.  The unit test waited for the tmux command to update the `status-right` line from the monitoring thread, but it never came, failing the unit test.

Possible Implementation
-----------------------

As mentioned, failed read events should in theory be triggered by something else again. So, it might be good to do something similiar to the following when [loading](https://github.com/BoniLindsley/phile/blob/6de42d3bb0337dca1c8295a73f45218304f6f3dc/phile/tray/gui.py#L124) to ignore errors:

```py
import json

def load(self):
    try:
        # First read the file contnet into a buffer.
        # This minimises the time spent on writing from the file.
        file_content = self.path.read_text()
        # And then parse the content.
        file_stream = io.StringIO(read_text())
        do_the_parsing_as_before_or_equivalent()
    except FileNotFoundError, json.decoder.JSONDecodeError:
        return
```

Also ignore file creation events in [`dispatch()`](https://github.com/BoniLindsley/phile/blob/master/phile/tray/tmux.py#L491) with say:

```py
if event_type == watchdog.events.EVENT_TYPE_CREATED:
    _logger.debug('Watchdog event: ignoring create events')
    return
if event_type == watchdog.events.EVENT_TYPE_MOVED:
    _logger.debug('Watchdog event: tray file moved.')
    for new_event_type in [
        watchdog.events.FileDeletedEvent,
        watchdog.events.FileModifiedEvent,
    ]:
        self.dispatch(new_event_type(watchdog_event.dest_path))
    return
```

I would prefer to not require cooperative locking.

Similar issues are likely to affect GUI code, and possibly notification file loading. It might be good to buffer file write as well, since the file sizes are expected to be small.

--- a/phile/tray/gui.py
+++ b/phile/tray/gui.py
@@ -2,14 +2,17 @@
 
 # Standard library.
 import bisect
+import io
 import json
 import logging
 import os
 import pathlib
 import pkg_resources
+import shutil
 import signal
 import sys
 import typing
+import warnings
 
 # External dependencies.
 from PySide2.QtCore import QObject
@@ -98,6 +101,12 @@ class TrayFile:
         """
         Parse tray file for a tray icon to be displayed.
 
+        :raises ~FileNotFoundError:
+            If the file to load cannot be found.
+        :raises ~json.decoder.JSONDecodeError:
+            If the JSON string part of the tray file to load
+            cannot be decode properly.
+
         The input file should start with a single line
         that can be displayed in a text tray environment such as `tmux`.
 
@@ -110,29 +119,28 @@ class TrayFile:
         It should not contan any other keys,
         and may be ignored, subject to implementation details.
         """
-
-        with self.path.open() as file_stream:
-
-            self.text_icon = file_stream.readline().rstrip('\r\n')
-            _logger.debug("Using text icon: %s", self.text_icon)
-
-            # Make sure there are content to read by reading one byte
-            # and then resetting the offset before decoding.
-            current_offset = file_stream.tell()
-            if file_stream.read(1):
-                file_stream.seek(current_offset)
-                json_content = json.load(file_stream)
-            else:
-                json_content = {}
-
-            self.icon_name = json_content.get('icon_name')
-            _logger.debug("Using icon name: %s", self.icon_name)
-            icon_path = json_content.get('icon_path')
-            if icon_path is not None:
-                self.icon_path = pathlib.Path(icon_path)
-            else:
-                self.icon_path = None
-            _logger.debug("Using icon path: %s", self.icon_path)
+        # Buffer the file content to reduce the chance of file changes
+        # introducing a race condition.
+        content_stream = io.StringIO(self.path.read_text())
+        # First line is the text icon.
+        # Do not write content yet in case JSON decoding fails.
+        text_icon = content_stream.readline().rstrip('\r\n')
+        # Make sure there are content to read by reading one byte
+        # and then resetting the offset before decoding.
+        current_offset = content_stream.tell()
+        if content_stream.read(1):
+            content_stream.seek(current_offset)
+            json_content = json.load(content_stream)
+        else:
+            json_content = {}
+        # Get properties from the decoded structure.
+        self.text_icon = text_icon
+        self.icon_name = json_content.get('icon_name')
+        icon_path = json_content.get('icon_path')
+        if icon_path is not None:
+            self.icon_path = pathlib.Path(icon_path)
+        else:
+            self.icon_path = None
 
     def remove(self):
         try:
@@ -141,19 +149,26 @@ class TrayFile:
             pass
 
     def save(self):
+        # Buffer for data to be written.
+        content_stream = io.StringIO()
+        # First line is the text icon.
+        if self.text_icon is not None:
+            content_stream.write(self.text_icon)
+        # Only copy over values that are filled in.
+        json_content: typing.Dict[str, str] = {}
+        for key in ['icon_name', 'icon_path']:
+            value = getattr(self, key, None)
+            if value is not None:
+                json_content[key] = str(value)
+        # If there is content to write, end the text icon line
+        # before writing the JSON string.
+        if json_content:
+            content_stream.write('\n')
+            json.dump(json_content, content_stream)
+        # Copy over the buffer.
         with self.path.open('w+') as file_stream:
-            if self.text_icon is not None:
-                file_stream.write(self.text_icon)
-            # End with a new line
-            # so that appending would not jumble up the text.
-            json_content: typing.Dict[str, str] = {}
-            for key in ['icon_name', 'icon_path']:
-                value = getattr(self, key, None)
-                if value is not None:
-                    json_content[key] = str(value)
-            if json_content:
-                file_stream.write('\n')
-                json.dump(json_content, file_stream)
+            content_stream.seek(0)
+            shutil.copyfileobj(content_stream, file_stream)
 
 
 def set_icon_paths() -> None:
@@ -282,7 +297,22 @@ class GuiIconList(QObject):
         # Determine what to do base on existence of the actual file.
         # There might be a delay between the file operation
         # and the event being received here.
-        self.load(tray_file)
+        # If the `load` fails, there is not much we can do about it.
+        # So just ignore.
+        # JSON string decoding error is likely to be
+        # a tray file modification in the middle of a read.
+        # In that case, a future file modification event is expected
+        # and this method will receive it, and we can handle it then.
+        # So ignoring is okay in that case.
+        try:
+            self.load(tray_file)
+        except json.decoder.JSONDecodeError:
+            warnings.warn(
+                'Unable to decode a tray file: {}'.format(
+                    tray_file.path
+                )
+            )
+            return
 
     def load(self, tray_file: TrayFile) -> None:
         # Figure out the position of the tray icon is in

--- a/phile/tray/tmux.py
+++ b/phile/tray/tmux.py
@@ -15,6 +15,7 @@ in control mode.
 # Standard libraries.
 import bisect
 import datetime
+import json
 import logging
 import os
 import pathlib
@@ -24,6 +25,7 @@ import shlex
 import subprocess
 import sys
 import typing
+import warnings
 
 # External dependencies.
 import watchdog.events  # type: ignore
@@ -528,7 +530,22 @@ class IconList(watchdog.events.FileSystemEventHandler):
         # Determine what to do base on existence of the actual file.
         # There might be a delay between the file operation
         # and the event being received here.
-        self.load(tray_file)
+        # If the `load` fails, there is not much we can do about it.
+        # So just ignore.
+        # JSON string decoding error is likely to be
+        # a tray file modification in the middle of a read.
+        # In that case, a future file modification event is expected
+        # and this method will receive it, and we can handle it then.
+        # So ignoring is okay in that case.
+        try:
+            self.load(tray_file)
+        except json.decoder.JSONDecodeError:
+            warnings.warn(
+                'Unable to decode a tray file: {}'.format(
+                    tray_file.path
+                )
+            )
+            return
 
     def load(self, tray_file: TrayFile) -> None:
         """

--- a/test_phile/test_tray/test_gui.py
+++ b/test_phile/test_tray/test_gui.py
@@ -411,6 +411,28 @@ class TestGuiIconList(unittest.TestCase):
         _logger.debug('End of test.')
 
     @unittest.mock.patch(
+        'phile.tray.gui.QSystemTrayIcon', SystemTrayIcon
+    )
+    def test_show_with_bad_file(self) -> None:
+        """
+        Show ignores badly structured files.
+
+        There is not much we can do about it as a reader.
+        """
+        # Create a tray.
+        tray = TrayFile(name='VeCat', configuration=self.configuration)
+        tray.icon_name = 'phile-tray-empty'
+        tray.text_icon = 'A'
+        tray.save()
+        with tray.path.open('a+') as file_stream:
+            file_stream.write('Extra text.')
+        # Check that they are all detected when showing the tray icons.
+        gui_icon_list = self.gui_icon_list
+        with self.assertWarns(UserWarning):
+            gui_icon_list.show()
+        self.assertEqual(len(gui_icon_list.tray_children()), 0)
+
+    @unittest.mock.patch(
         'phile.tray.gui.QIcon.fromTheme', q_icon_from_theme
     )
     @unittest.mock.patch(

--- a/test_phile/test_tray/test_tmux.py
+++ b/test_phile/test_tray/test_tmux.py
@@ -499,6 +499,24 @@ class TestIconList(unittest.TestCase):
         self.icon_list.load(year_tray_file)
         self.control_mode.send_command.assert_not_called()
 
+    def test_dispatch_with_bad_file(self) -> None:
+        """
+        Dispatch ignores badly structured files.
+
+        There is not much we can do about it as a reader.
+        """
+        tray_file = TrayFile(
+            configuration=self.configuration, name='month'
+        )
+        tray_file.text_icon = 'ABC'
+        tray_file.save()
+        with tray_file.path.open('a+') as file_stream:
+            file_stream.write('\nNot JSON.')
+        with self.assertWarns(UserWarning):
+            self.icon_list.dispatch(
+                watchdog.events.FileCreatedEvent(tray_file.path)
+            )
+
     def test_show_with_existing_files(self) -> None:
         """
         Showing should display existing tray files.

